File: src/Prism.Core/Syntax/Green/GreenTriviaList.cs
// @file GreenTriviaList.cs
//
// @copyright Copyright (c) 2026 Retro & Chill. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.

namespace Prism.Core.Syntax.Green;

internal sealed class GreenTriviaList : GreenNode
{
    private readonly GreenTrivia[] _trivia;

    public static readonly GreenTriviaList Empty = new([]);

    public bool IsEmpty => _trivia.Length == 0;

    public GreenTriviaList(GreenTrivia[] trivia)
        : base(SyntaxKind.TriviaList, CalculateWidth(trivia))
    {
        _trivia = trivia;
        ChildCount = _trivia.Length;
    }

    private static int CalculateWidth(GreenTrivia[] trivia)
    {
        return trivia.Sum(t => t.FullWidth);
    }

    public override GreenTrivia? GetChild(int index)
    {
        return index < _trivia.Length ? _trivia[index] : null;
    }
}

File: libs/core/src/private/syntax/green/green_trivia_list.cpp
/**
 * @file green_trivia_list.cpp
 * @author Francesco Corso
 * @date 7/9/2026
 * @brief
 */
module prism.core:syntax.green.green_trivia_list.impl;

import :syntax.green.green_trivia_list;

namespace prism
{
    std::uint32_t GreenTriviaList::calculate_width(const std::vector<GreenPtr<GreenTrivia>> &trivia)
    {
        return std::ranges::fold_left(trivia,
                                      0u,
                                      [](const std::uint32_t acc, const auto &t) { return acc + t->full_width(); });
    }

    GreenPtr<GreenTriviaList> normalize_trivia(GreenPtr<GreenTriviaList> trivia)
    {
        if (trivia == nullptr || trivia->empty())
            return nullptr;

        return std::move(trivia);
    }
} // namespace prism
